src/emulator/extensionsEmulator.ts

```diff
@@ -67,10 +70,20 @@ export class ExtensionsEmulator {
       path.join(os.homedir(), ".cache", "firebase", "extensions");
     const sourceCodePath = path.join(cacheDir, ref);
 
+    // Wait 20s if the source with same ref is being downloaded already before we check source validity.
```

Comment:
This 20s wait is very ugly. Instead, i think we can make pendingDownloads a map of ref -> download promise. Then, if pendingDownloads[ref] exists, we can  just await it

Reply:
ok I agree the solution is hacky lol, updated based on your suggestion